craft_application/models/grammar.py

```diff
@@ -52,6 +52,19 @@ class GrammarAwareProject(_GrammarAwareModel):
 
     parts: "dict[str, _GrammarAwarePart]"
 
+    @pydantic.root_validator(  # pyright: ignore[reportUntypedFunctionDecorator,reportUnknownMemberType]
+        pre=True
+    )
+    def _ensure_parts(cls, data: dict[str, Any]) -> dict[str, Any]:
```

Comment:
out of curiosity, couldn't we do `parts: "dict[str, _GrammarAwarePart]" = {}`? Or non Field equivalent

Reply:
We probably could, but I chose to do it this way because I'm not sure we want to support this behaviour forever. Ideally, a future charmcraft would begin to require parts to be specified.